streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java

```diff
@@ -46,6 +46,79 @@ RocksDBStore getRocksDBStore() {
         return new RocksDBTimestampedStore(DB_NAME);
     }
 
+    @Test
+    public void shouldOpenNewStoreInRegularMode() {
+        LogCaptureAppender.setClassLoggerToDebug(RocksDBTimestampedStore.class);
+
+        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+        rocksDBStore.init(context, rocksDBStore);
+        assertThat(appender.getMessages(), hasItem("Opening store " + DB_NAME + " in regular mode"));
+        LogCaptureAppender.unregister(appender);
+
+        try (final KeyValueIterator<Bytes, byte[]> iterator = rocksDBStore.all()) {
+            assertThat(iterator.hasNext(), is(false));
+        }
+    }
+
+    @Test
+    public void shouldOpenExistingStoreInRegularMode() throws Exception {
+        LogCaptureAppender.setClassLoggerToDebug(RocksDBTimestampedStore.class);
+
+        // prepare store
+        rocksDBStore.init(context, rocksDBStore);
+        rocksDBStore.put(new Bytes("key".getBytes()), "timestamped".getBytes());
+        rocksDBStore.close();
+
+        // re-open store
+        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+        rocksDBStore = getRocksDBStore();
+        rocksDBStore.init(context, rocksDBStore);
+        assertThat(appender.getMessages(), hasItem("Opening store " + DB_NAME + " in regular mode"));
+        LogCaptureAppender.unregister(appender);
+
+        rocksDBStore.close();
+
+        // verify store
+        final DBOptions dbOptions = new DBOptions();
+        final ColumnFamilyOptions columnFamilyOptions = new ColumnFamilyOptions();
+
+        final List<ColumnFamilyDescriptor> columnFamilyDescriptors = asList(
+            new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
+            new ColumnFamilyDescriptor("keyValueWithTimestamp".getBytes(StandardCharsets.UTF_8), columnFamilyOptions));
+        final List<ColumnFamilyHandle> columnFamilies = new ArrayList<>(columnFamilyDescriptors.size());
+
+        RocksDB db = null;
+        ColumnFamilyHandle noTimestampColumnFamily = null, withTimestampColumnFamily = null;
+        try {
+            db = RocksDB.open(
+                dbOptions,
+                new File(new File(context.stateDir(), "rocksdb"), DB_NAME).getAbsolutePath(),
+                columnFamilyDescriptors,
+                columnFamilies);
+
+            noTimestampColumnFamily = columnFamilies.get(0);
+            withTimestampColumnFamily = columnFamilies.get(1);
+
+            assertThat(db.get(noTimestampColumnFamily, "key".getBytes()), new IsNull<>());
+            assertThat(db.getLongProperty(noTimestampColumnFamily, "rocksdb.estimate-num-keys"), is(0L));
```

Comment:
Do we know that `rocksdb.estimate-num-keys` would always return 0 and 1 in this test (since it is only an approximate estimate, hence no guarantee)?

Reply:
It's no exact counting in general, but this this case we know it will be counting correctly However, the "approximation" is at least deterministic and hence, it seems safe to me to write the test that way.

Basically, for each put() RocksDB increments the counter and for each delete() is decrements the counter. Hence, if you do a put() for an existing key, RockDB over counts, and if you do a delete() for a non-exiting key, RocksDB under counts.

In this test, we do a single put() on one CF and not operation on the other CF, hence, the count (ie, approximation) will be correct.

Thoughts?

Reply:
SG.